content/en/docs/concepts/cluster-administration/node-autoscaling.md

```diff
@@ -0,0 +1,264 @@
+to optimize different properties (for example the maximum lifespan of Nodes in a
+cluster).
+
+The details of how consolidation is performed depend on the configuration of a
+given autoscaler. Consult [Configure workloads for autoscaling](#workload-config)
+for common workload-level configuration options.
+
+{{< note >}}
+Consolidation is formerly known as _scale-down_ in Cluster Autoscaler.
+{{< /note >}}
+
+## Autoscalers {#autoscalers}
+
+The functionalities described in previous sections are provided by Node _autoscalers_.
+In addition to the Kubernetes API, autoscalers also need to interact with cloud provider
+APIs to provision and consolidate Nodes. This means that they need to be explicitly
+integrated with each supported cloud provider. The performance and feature set of
+a given autoscaler can differ between cloud provider integrations.
+
+{{< mermaid >}}
+graph TD
+    na[Node autoscaler]
+    k8s[Kubernetes]
+    cp[Cloud Provider]
+
+    k8s --> |get Pods/Nodes|na
+    na --> |drain Nodes|k8s
+    na --> |create/remove VMs|cp
+    cp --> |get VMs|na
+
+    classDef white_on_blue fill:#326ce5,stroke:#fff,stroke-width:4px,color:#fff;
+    classDef blue_on_white fill:#fff,stroke:#bbb,stroke-width:2px,color:#326ce5;
+    class na blue_on_white;
+    class k8s,cp white_on_blue;
+{{</ mermaid >}}
+
+### Autoscaler implementations
+
+Cluster autoscaler and Karpenter are the two Node autoscalers currently supported by
+[sig-autoscaling](https://github.com/kubernetes/community/tree/master/sig-autoscaling).
+
+From the perspective of a cluster user, both autoscalers should provide a similar Node autoscaling
+experience. Both will provision new Nodes for unschedulable Pods, and both will
+consolidate the Nodes that are no longer optimally utilized.
+
+The most of the difference between the autoscalers is in how they're set up and
+configured, as well as in how well they're integrated with a given cloud provider.
+The autoscalers also provide features outside the Node autoscaling scope described
+on this page, and the features differ between them.
+
+Consult the sections below, and the linked documentation for the individual
+autoscalers to decide which autoscaler fits your use-case better.
+
+#### Cluster Autoscaler
+
+Cluster Autoscaler adds or removes Nodes to pre-configured Node groups.
+Node groups generally map to some sort of provider-side VM group. A
+single instance of Cluster Autoscaler can simultaneously manage multiple
+Node groups. When provisioning, Cluster Autoscaler will add Nodes to the group
+that best fits the requests of pending Pods. When consolidating, Cluster Autoscaler
+always selects specific Nodes to remove as opposed to just resizing the underlying VM group.
+
+Main differences vs Karpenter:
+* Cluster Autoscaler (CA) provides features related to just Node autoscaling. Node lifecycle
+  beyond that is not in its scope.
+* CA doesn't support auto-provisioning, the Node groups it can provision from have to be
```

Comment:
Rather than "auto-provisioning" maybe mixed sku autoscaling makes more sense.  

We could also do a small writeup of the advantages of mixed sku autoscaling. Here are some examples that came to my mind when trying to write up some advantages to having access to all skus. Someone can perhaps word it better than me.
----------

1. "75% of scale up failures for Azure's Cluster Autoscaler come from lack of configured quota for the 4 or so sizes customers configure, with less than 1% of customers configuring CAS with more than 4 vm sizes on aks." With mixed sku autoscaling, you aren't limited to those 4 vm sizes. 

2. The per nodeclaim model vs the per nodegroup model allows for different flexibility in the retry model. 

2a. Karpenter puts the responsibility on the cloud provider to select the SKU. Meaning it can choose to use factors like zonal availability, quota, price, and allocation errors in its VM SKU selection decision. 

2b. CAS can't self mitigate the quota errors seen in scale up, but karpenter can retry with a different SKU that has quota or a different zone if there are zonal allocation failures. So retrying continuously in karpenter is less likely to fail.  

These are some of the benefits of having a mixed SKU model, and a per vm retry model. We can retry more intelligently caching skus as unavailable due to zonal, or quota related reasons. 

Another large factor to consider is the pricing selection, being given all SKUs on a given cloudprovider, allows for much better cost savings. CAS has the pricing expander, but the real weakness is that it doesn't have 1000 sizes to pick from. It just can pick based on the current VM Sizes configured by the customer. (As stated earlier, is often no more than 4)

Reply:
great details @Bryce-Soghigian , i'm not so comfortable about using "SKU" so much, as many people are using on-prem and bare metal deployments which look different from public cloud offerings. that said, i totally get what you are talking about.

@jackfrancis this is a tough one for me, because at the core, CAS needs to be informed about the pools it can scale, whereas karpenter can be given some recipes on how to find instances to create pools. scale from zero might appear like autoprovisioning, but the big difference for me is that you still need to configure the node group and link it to an instance type. in cluster-api the user would need to create all the various infrastructure machine templates to be able to have those instance types represented. but on karpenter that expression would be much more compact.

for me, the part to clarify is that CAS will only be able to scale groups that it knows about but karpenter can discover new instances by the constraints it is given. assuming we want to clarify this more.

perhaps we could use a section that describes the difference between "scaling" as opposed to "provisioning" ?

Reply:
IIUC, it seems the core difference between the two instance-shape calculations is that Karpenter has more attribute based selection features, and CAS does more instance type matching. Karpenter's NodePools are defaulted to fully inclusive (default = *) where IIUC, CAS is defaulted to fully exclusive, where no instance types in a node group will not be compatible with any pod.

Reply:
> CAS has the pricing expander, but the real weakness is that it doesn't have 1000 sizes to pick from.

In GKE (which is where I have the data on) we see many clusters with >100 nodepools (GKE nodepools; they can be thought of as equivalent of CAS nodegroups). That's enough to effectively give you full spectrum of shapes for optimization (GCE has less instance types than AWS; 100-150 is enough to cover full spectrum of general purpose instances). So even with CAS you can have the SKU selection and zonal handling. 

A major difference is that you need to go through the process of setting all those nodegroups and this needs to be done outside of CAS. With auto-provisioning you no longer have to do any external setup and you get all the different shapes with just a simple configuration of autoscaler.

Reply:
Yep that makes a lot of sense. My point was more from the azure perspective we do not see many clusters creating that many nodepools/nodegroups. GKE has a wonderful CAS integration the rest of us need to strive towards. 

I would also argue at a certain point having 1000 vms that all do the same thing isn't a strength and that 150 or so sizes is more than enough for all intents and purposes.

Reply:
Judging by this discussion, I think we should probably expand the "Node auto-provisioning" section under "Provisioning", and then this part will be clear?

@njtran's analogy is interesting, but I'm not sure if it maps onto CAS. I see the 2 modes like this, but perhaps I'm missing some Karpenter context:

* Regular provisioning: The user defines a set of homogenous cloud provider VM groups. The groups allow for adding and removing VMs. Based on the pending pods, autoscaler chooses two parameters for provisioning: which group, how many VMs to add. For the autoscaler, this is a well-scoped problem: among the VM options that the user pre-configured, figure out how many new VMs would be needed, then choose the "best" <group, node_count> option, and provision that many Nodes from that group.
* Auto-provisioning: The user doesn't have to define an explicit cloud provider VM group for every possible configuration that could be needed for their workloads. The user only defines a set of constraints on the VMs that can be provisioned. Based on the pending pods, and on the set of constraints, the autoscaler chooses two parameters: the VM config, how many VMs to add. Choosing a VM config from the vast space of possible cloud provider configs is IMO a fundamentally different problem than just selecting between pre-configured configs.

> One could argue that "scale from zero" is functionally equivalent to "auto-provisioning".

@jackfrancis I would disagree here, there's no reason why the regular provisioning model I described above shouldn't work with scale-from-zero. Even while scaling from >0, the autoscaler has to predict how a new Node would look like if a new VM was added to a given group. This prediction is usually straightforward if there are existing Nodes in the group - just take the existing Node and sanitize it. If there are no Nodes in a given group, the prediction has to come from the cloud provider. So some cloud provider integrations will support it while some others won't, but IMO this is more of an implementation detail of a given cloud provider integration for regular provisioning.

From another perspective, the only thing that the autoscaler has to solve on top of scale-from-not-zero to have scale-from-zero is passing "VM template" information from the cloud provider VM group. This information should be available in the group in order to create new VMs, so this should only be the matter of some plumbing and then it's the same as from-not-zero. Contrast this with auto-provisioning where there's no pre-configured template that can be leveraged.